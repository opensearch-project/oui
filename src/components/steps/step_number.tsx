/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import classNames from 'classnames';
import React, { FunctionComponent, HTMLAttributes } from 'react';
import { OuiScreenReaderOnly } from '../accessibility';
import { CommonProps, keysOf } from '../common';
import { OuiIcon } from '../icon';
import { OuiStepProps } from './step';
import {
  useI18nCompleteStep,
  useI18nDisabledStep,
  useI18nErrorsStep,
  useI18nIncompleteStep,
  useI18nStep,
  useI18nWarningStep,
  useI18nLoadingStep,
} from './step_strings';
import { OuiLoadingSpinner } from '../loading';

const statusToClassNameMap = {
  incomplete: 'ouiStepNumber--incomplete',
  disabled: 'ouiStepNumber--disabled',
  loading: 'ouiStepNumber--loading',
  warning: 'ouiStepNumber--warning',
  danger: 'ouiStepNumber--danger',
  complete: 'ouiStepNumber--complete',
};

export const STATUS = keysOf(statusToClassNameMap);
export type OuiStepStatus = typeof STATUS[number];

export interface OuiStepNumberProps
  extends CommonProps,
    HTMLAttributes<HTMLDivElement> {
  /**
   * May replace the number provided in props.number with alternate styling
   */
  status?: OuiStepStatus;
  number?: number;
  /**
   * Uses a border and removes the step number.
   */
  isHollow?: boolean;
  /**
   * Title sizing equivalent to OuiTitle, but only `m`, `s` and `xs`. Defaults to `s`
   */
  titleSize?: OuiStepProps['titleSize'];
}

export const OuiStepNumber: FunctionComponent<OuiStepNumberProps> = ({
  className,
  status,
  number,
  isHollow,
  titleSize,
  ...rest
}) => {
  const stepAriaLabel = useI18nStep({ number });
  const completeAriaLabel = useI18nCompleteStep({ number });
  const warningAriaLabel = useI18nWarningStep({ number });
  const errorsAriaLabel = useI18nErrorsStep({ number });
  const incompleteAriaLabel = useI18nIncompleteStep({ number });
  const disabledAriaLabel = useI18nDisabledStep({ number });
  const loadingAriaLabel = useI18nLoadingStep({ number });

  const classes = classNames(
    'ouiStepNumber',
    status ? statusToClassNameMap[status] : undefined,
    { 'ouiStepNumber-isHollow': isHollow },
    className
  );

  const iconSize = titleSize === 'xs' ? 's' : 'm';
  let screenReaderText = stepAriaLabel;
  if (status === 'incomplete') screenReaderText = incompleteAriaLabel;
  else if (status === 'disabled') screenReaderText = disabledAriaLabel;
  else if (status === 'loading') screenReaderText = loadingAriaLabel;

  let numberOrIcon = (
    <>
      <OuiScreenReaderOnly>
        <span>{screenReaderText}</span>
      </OuiScreenReaderOnly>
      <span className="ouiStepNumber__number" aria-hidden="true">
        {number}
      </span>
    </>
  );

  if (status === 'complete') {
    numberOrIcon = (
      <OuiIcon
        type="check"
        className="ouiStepNumber__icon"
        size={iconSize}
        aria-label={completeAriaLabel}
      />
    );
  } else if (status === 'warning') {
    numberOrIcon = (
      <OuiIcon
        type="alert"
        className="ouiStepNumber__icon"
        size={iconSize}
        aria-label={warningAriaLabel}
      />
    );
  } else if (status === 'danger') {
    numberOrIcon = (
      <OuiIcon
        type="cross"
        className="ouiStepNumber__icon"
        size={iconSize}
        aria-label={errorsAriaLabel}
      />
    );
  } else if (status === 'loading') {
    numberOrIcon = (
      <>
        <OuiScreenReaderOnly>
          <span>{screenReaderText}</span>
        </OuiScreenReaderOnly>
        <OuiLoadingSpinner
          className="ouiStepNumber__loader"
          size={iconSize === 's' ? 'l' : 'xl'}
        />
      </>
    );
  }

  return (
    <span className={classes} {...rest}>
      {numberOrIcon}
    </span>
  );
};
