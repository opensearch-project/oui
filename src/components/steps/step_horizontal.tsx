/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import classNames from 'classnames';
import React, {
  ButtonHTMLAttributes,
  FunctionComponent,
  MouseEvent as ReactMouseEvent,
  MouseEventHandler,
} from 'react';
import { CommonProps } from '../common';
import { OuiStepNumber, OuiStepStatus } from './step_number';
import {
  useI18nCompleteStep,
  useI18nDisabledStep,
  useI18nIncompleteStep,
  useI18nStep,
  useI18nWarningStep,
} from './step_strings';

export interface OuiStepHorizontalProps
  extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'onClick'>,
    CommonProps {
  /**
   * Adds to the line before the indicator for showing current progress
   */
  isSelected?: boolean;
  /**
   * Adds to the line after the indicator for showing current progress
   */
  isComplete?: boolean;
  onClick: MouseEventHandler<HTMLButtonElement>;
  /**
   * Makes the whole step button disabled.
   */
  disabled?: boolean;
  /**
   * The number of the step in the list of steps
   */
  step?: number;
  title?: string;
  /**
   * Visual representation of the step number indicator.
   * May replace the number provided in props.step with alternate styling.
   * The `isSelected`, `isComplete`, and `disabled` props will override these.
   */
  status?: OuiStepStatus;
}

export const OuiStepHorizontal: FunctionComponent<OuiStepHorizontalProps> = ({
  className,
  step = 1,
  title,
  isSelected,
  isComplete,
  onClick,
  disabled,
  status,
  ...rest
}) => {
  const buttonTitle = useI18nStep({ number: step, title });
  const completeTitle = useI18nCompleteStep({ number: step, title });
  const disabledTitle = useI18nDisabledStep({ number: step, title });
  const incompleteTitle = useI18nIncompleteStep({ number: step, title });
  const warningTitle = useI18nWarningStep({ number: step, title });

  const classes = classNames('ouiStepHorizontal', className, {
    'ouiStepHorizontal-isSelected': isSelected,
    'ouiStepHorizontal-isComplete': isComplete,
    'ouiStepHorizontal-isIncomplete': !isSelected && !isComplete,
    'ouiStepHorizontal-isDisabled': disabled,
  });

  if (disabled) status = 'disabled';
  else if (isComplete) status = 'complete';
  else if (isSelected) status = status;
  else if (!status) status = 'incomplete';

  let stepTitle = buttonTitle;
  if (status === 'disabled') stepTitle = disabledTitle;
  if (status === 'complete') stepTitle = completeTitle;
  if (status === 'incomplete') stepTitle = incompleteTitle;
  if (status === 'warning') stepTitle = warningTitle;

  const onStepClick = (
    event: ReactMouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    if (!disabled) onClick(event);
  };

  return (
    <button
      className={classes}
      title={stepTitle}
      onClick={onStepClick}
      disabled={disabled}
      {...rest}>
      <OuiStepNumber
        className="ouiStepHorizontal__number"
        status={status}
        number={step}
      />

      <span className="ouiStepHorizontal__title">{title}</span>
    </button>
  );
};
