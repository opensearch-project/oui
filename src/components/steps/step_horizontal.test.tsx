/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import { requiredProps } from '../../test/required_props';

import { STATUS } from './step_number';
import { OuiStepHorizontal } from './step_horizontal';

describe('OuiStepHorizontal', () => {
  test('is rendered', () => {
    const { container } = render(
      <OuiStepHorizontal {...requiredProps} onClick={() => {}} />
    );

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    test('step', () => {
      const { container } = render(
        <OuiStepHorizontal step={5} onClick={() => {}} />
      );

      expect(container).toMatchSnapshot();
    });

    test('title', () => {
      const { container } = render(
        <OuiStepHorizontal title={'First step'} onClick={() => {}} />
      );

      expect(container).toMatchSnapshot();
    });

    test('isSelected', () => {
      const { container } = render(
        <OuiStepHorizontal isSelected onClick={() => {}} />
      );

      expect(container).toMatchSnapshot();
    });

    test('isComplete', () => {
      const { container } = render(
        <OuiStepHorizontal isComplete onClick={() => {}} />
      );

      expect(container).toMatchSnapshot();
    });

    describe('status', () => {
      STATUS.forEach((status) => {
        test(`${status} is rendered`, () => {
          const { container } = render(
            <OuiStepHorizontal status={status} onClick={() => {}} />
          );

          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('onClick', () => {
      test('is called when clicked', () => {
        const onClickHandler = jest.fn();

        render(<OuiStepHorizontal step={1} onClick={onClickHandler} />);

        act(() => {
          fireEvent.click(screen.getByRole('button'));
        });

        expect(onClickHandler).toHaveBeenCalledTimes(1);
      });

      test("isn't called when clicked if it's disabled", () => {
        const onClickHandler = jest.fn();

        render(
          <OuiStepHorizontal disabled step={1} onClick={onClickHandler} />
        );

        act(() => {
          fireEvent.click(screen.getByRole('button'));
        });

        expect(onClickHandler).not.toHaveBeenCalled();
      });
    });
  });
});
