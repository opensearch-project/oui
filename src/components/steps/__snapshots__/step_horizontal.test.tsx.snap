// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiStepHorizontal is rendered 1`] = `
<div>
  <button
    aria-label="aria-label"
    class="ouiStepHorizontal testClass1 testClass2 ouiStepHorizontal-isIncomplete"
    data-test-subj="test subject string"
    title="Step 1 is incomplete"
  >
    <span
      class="ouiStepNumber ouiStepNumber--incomplete ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is incomplete
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props isComplete 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isComplete"
    title="Step 1 is complete"
  >
    <span
      class="ouiStepNumber ouiStepNumber--complete ouiStepHorizontal__number"
    >
      <span
        aria-label="Step 1 is complete"
        class="ouiStepNumber__icon"
        data-ouiicon-type="check"
      />
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props isSelected 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isSelected"
    title="Step 1"
  >
    <span
      class="ouiStepNumber ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props status complete is rendered 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1 is complete"
  >
    <span
      class="ouiStepNumber ouiStepNumber--complete ouiStepHorizontal__number"
    >
      <span
        aria-label="Step 1 is complete"
        class="ouiStepNumber__icon"
        data-ouiicon-type="check"
      />
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props status danger is rendered 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1"
  >
    <span
      class="ouiStepNumber ouiStepNumber--danger ouiStepHorizontal__number"
    >
      <span
        aria-label="Step 1 has errors"
        class="ouiStepNumber__icon"
        data-ouiicon-type="cross"
      />
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props status disabled is rendered 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1 is disabled"
  >
    <span
      class="ouiStepNumber ouiStepNumber--disabled ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is disabled
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props status incomplete is rendered 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1 is incomplete"
  >
    <span
      class="ouiStepNumber ouiStepNumber--incomplete ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is incomplete
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props status loading is rendered 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1"
  >
    <span
      class="ouiStepNumber ouiStepNumber--loading ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is loading
      </span>
      <span
        class="ouiLoadingSpinner ouiLoadingSpinner--xLarge ouiStepNumber__loader"
      />
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props status warning is rendered 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1 has warnings"
  >
    <span
      class="ouiStepNumber ouiStepNumber--warning ouiStepHorizontal__number"
    >
      <span
        aria-label="Step 1 has warnings"
        class="ouiStepNumber__icon"
        data-ouiicon-type="alert"
      />
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props step 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 5 is incomplete"
  >
    <span
      class="ouiStepNumber ouiStepNumber--incomplete ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 5 is incomplete
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        5
      </span>
    </span>
    <span
      class="ouiStepHorizontal__title"
    />
  </button>
</div>
`;

exports[`OuiStepHorizontal props title 1`] = `
<div>
  <button
    class="ouiStepHorizontal ouiStepHorizontal-isIncomplete"
    title="Step 1: First step is incomplete"
  >
    <span
      class="ouiStepNumber ouiStepNumber--incomplete ouiStepHorizontal__number"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is incomplete
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <span
      class="ouiStepHorizontal__title"
    >
      First step
    </span>
  </button>
</div>
`;
